File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    ListNode* reverseList(ListNode* head) {
        ListNode *prev = nullptr;
        while (head) {
            ListNode* next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
    ListNode* getKthNode(ListNode* temp, int k) {
        for (int i=0; temp != nullptr && i < k - 1; ++i) {
            temp = temp->next;
        }
        return temp;
    }
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* curr = dummy;
        while (head) {
            ListNode* kthNode = getKthNode(head, k);
            if (!kthNode) break;
            ListNode* nextGroup = kthNode->next;
            kthNode->next = nullptr;
            curr->next = reverseList(head);
            curr = head; // head is the tail after reverse for any group cause we using head even below look
            curr->next = nextGroup;
            head = nextGroup; // head is gonna be tail for reverse of this group
        }
        return dummy->next;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
class Solution {
    private ListNode reverseList(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode prev = null;
        ListNode current = head;
        while (current != null) {
            ListNode follow = current.next;
            current.next = prev;
            prev = current;
            current = follow;
        }
        return prev;
    }

    private static ListNode getTail(ListNode temp, int k) {
        for (int i = 0; temp!=null & i<k-1; ++i)
            temp = temp.next;
        return temp;
    }

    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prevLast = dummy;

        while (head != null) {
            ListNode tail = getTail(head, k);
            if (tail == null) {
                break;
            }
            ListNode nextGroup = tail.next;
            tail.next = null;
            prevLast.next = reverseList(head);
            prevLast = head;
            head.next = nextGroup;
            head = nextGroup;
        }

        return dummy.next;
    }
}
